### PR TITLE
Converted prefer-declare-function from TSLint to ESLint

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
Continues the work from #489.

### CODE REVIEW

**Comment by JoshuaKGoldberg** [comment]
Bump @andrewbranch - is there any chance this & the other open TSLint->ESLint PRs be looked at soon, please?

**Comment by sandersn** [comment]
@JoshuaKGoldberg I work on these during my DT week but I don't think anybody else is. When everybody gets back I'll ask others to take time as well.

File: packages/dtslint/src/rules/prefer-declare-function.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+      variableFunction: "Use a function declaration instead of a variable of function type.",
+    },
+    schema: [],
+    type: "problem",
+  },
+  name: "prefer-declare-function",
+  create(context) {
+    return {
+      // eslint-disable-next-line @typescript-eslint/naming-convention
+      "ExportNamedDeclaration > VariableDeclaration > VariableDeclarator"(node: TSESTree.VariableDeclarator) {
```
Comment (sandersn) [inline packages/dtslint/src/rules/prefer-declare-function.ts:22]: the type doesn't need to be exported to be an error.

**Review by sandersn — CHANGES_REQUESTED** [review]
The rule should apply to all variable declarations, not just exported ones.
I'll make the change since it's a simple one and this PR has been unreviewed for a long time.

File: packages/dtslint/src/rules/preferDeclareFunctionRule.ts [hunk 2]
```diff
@@ -1,36 +0,0 @@
...
-    "Use a function declaration instead of a variable of function type."
-  );
-
-  apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
-    return this.applyWithFunction(sourceFile, walk);
-  }
-}
-
-function walk(ctx: Lint.WalkContext<void>): void {
-  eachModuleStatement(ctx.sourceFile, (statement) => {
```
Comment (sandersn) [inline packages/dtslint/src/rules/preferDeclareFunctionRule.ts:27]: I missed this call to eachModuleStatement. It

1. Only runs on statements in declaration files.
2. Recursively runs on `declare module "foo"` blocks. This includes namespaces too, as far as I can tell.

The selector on line 22 is probably good enough for (2), but I'm not sure how to make (1) happen yet.

**Review by sandersn — APPROVED** [review]

**Comment by sandersn** [comment]
I'll ship this tomorrow so I can create my fixup PR on DT right after.

**Pull request merged**